src/librbd/image_copy/ObjectCopyRequest.cc

```diff
@@ -0,0 +1,632 @@
+    uint64_t dst_object_off = off % dst_object_size;
+    size_t len = std::min(src_object_size - src_object_off,
+                          dst_object_size - dst_object_off);
+
+    m_src_object_offsets[src_object_number] = {src_object_off, len};
+    off += len;
+  }
+  assert(off == dst_object_size * (m_dst_object_number + 1));
+  assert(!m_src_object_offsets.empty());
+
+  ldout(m_cct, 20) << m_src_object_offsets.size() << " src objects, first: "
+                   << *m_src_object_offsets.begin() << dendl;
+}
+
+template <typename I>
+void ObjectCopyRequest<I>::compute_diffs() {
+  auto it = m_src_object_offsets.rbegin();
+  assert(it != m_src_object_offsets.rend());
+  auto src_object_number = it->first;
+  auto src_object_off = it->second.first;
+  auto len = it->second.second;
+
+  interval_set<uint64_t> copy_interval;
+  copy_interval.insert(src_object_off, src_object_off + len);
+
+  bool first_src_object = (m_src_object_offsets.size() == 1);
+
+  librados::snap_t src_copy_point_snap_id = m_snap_map.rbegin()->first;
+  uint64_t prev_end_size = 0;
+  bool prev_exists = false;
+  librados::snap_t start_src_snap_id = 0;
+  for (auto &pair : m_snap_map) {
+    assert(!pair.second.empty());
+    librados::snap_t end_src_snap_id = pair.first;
+    librados::snap_t end_dst_snap_id = pair.second.front();
+
+    interval_set<uint64_t> diff;
+    uint64_t end_size;
+    bool exists;
+    librados::snap_t clone_end_snap_id;
+    calc_snap_set_diff(m_cct, m_snap_set, start_src_snap_id,
+                       end_src_snap_id, &diff, &end_size, &exists,
+                       &clone_end_snap_id);
+
+    ldout(m_cct, 20) << "start_src_snap_id=" << start_src_snap_id << ", "
+                     << "end_src_snap_id=" << end_src_snap_id << ", "
+                     << "clone_end_snap_id=" << clone_end_snap_id << ", "
+                     << "end_dst_snap_id=" << end_dst_snap_id << ", "
+                     << "diff=" << diff << ", "
+                     << "end_size=" << end_size << ", "
+                     << "exists=" << exists << dendl;
+
+    bool allow_trunc = m_snap_object_states.find(end_dst_snap_id) ==
+      m_snap_object_states.end();
+
+    // limit diff to copy interval
+    diff.intersection_of(copy_interval);
+    ldout(m_cct, 20) << "copy diff: " << diff << dendl;
+    if (end_size < src_object_off) {
+      end_size = src_object_off;
+    } else if (end_size > src_object_off + len) {
+      end_size = src_object_off + len;
+    }
+
+    if (exists) {
+      // clip diff to size of object (in case it was truncated)
+      if (end_size < prev_end_size) {
+        interval_set<uint64_t> trunc;
+        trunc.insert(end_size, prev_end_size);
+        trunc.intersection_of(diff);
+        diff.subtract(trunc);
+        ldout(m_cct, 20) << "clearing truncate diff: " << trunc << dendl;
+      }
+
+      // prepare the object map state
+      // XXXMG: how to support OBJECT_EXISTS_CLEAN?
```

Comment:
@dillaman What do you think about this? The commented code is from the original image_sync. Now, thinking about this more it looks to me I can safely use the same, i.e. just uncomment this code and remove `m_snap_object_states[end_dst_snap_id] = OBJECT_EXISTS`.

Reply:
Seems OK to me